### DISCUSSION TITLE
How to convert polygon to H3 cells where the cells completely cover the shape ?

### DISCUSSION THREAD

**Discussion by michaelbrichko** [uber/h3#798] [General]
Hi,

The functionality that library provides is to approximate polygon shape with H3 cells where every cell center resides inside the shape but what I want is to cover the shape completely with H3 cells such that I have a minimal set of H3 cells which cover the shape completely.

The option that I can think of is given H3 cells approximation, add their immidiete neighbors to the result.
I have two issues with that:
1. Some neigbors are completely doesn't intersect with the shape, so they are absolutely redundant.
2. It's much more compute

Is there a better option to solve this?
Thanks!!!

**Comment by isaacbrodsky** [comment]
This is being actively worked on in #796, as the `OVERLAPPING` mode. There are workarounds involving buffering the H3 cells from the current algorithm but it sounds like you have already explored most of those options.

**Reply by michaelbrichko** [reply]
Amazing! Thank you!
Is there a Buffer functionality as well on the horizon ? I mean, I would like to have the possibility to find overlapping cells of polygon+buffer around that polygon :)

**Reply by nrabinowitz** [reply]
We've thought about this for a while, but the fact is, we haven't thought of a way to do it that wouldn't be equally fast (or faster) using the native Set class in your target language. The basic approach (in Python):
```
hex_set = set(cells)
for cell in hex_set:
  set.update(h3.grid_disk(cell, radius))
```
While we have a relatively efficient approach for creating a set of H3 indexes in C, in general we've felt that set operations wouldn't benefit from being implemented in C, due to the issues of memory management and the costs of moving data between bindings and C, and end users would likely be better off using the native set operations in their language of choice.

**Reply by michaelbrichko** [reply]
Thank you! Let me clarify about the buffer ask please, is there a plan to support overlapping of cells with polygon+buffer where buffer is distance in meters from the polygon border ?

**Reply by michaelbrichko** [reply]
A follow up question please. With already existing tools, how would you suggest to calculate how many more rings to add to polygon in order to be sure that eventual set of cells overlap with polygon+radius in meters. I suspect the main part of solving this is with how many cells the radius in meters should be covered? Given specific resolution, I can calculate hexagon edge length. Deviding buffer length by hexagon edge should be it. But this is a bit combersome, plus there is a chance to stumble on pentagon. What would be better way to solve this ?

Thanks you!

**Reply by michaelbrichko** [reply]
Or perhaps somehow generate 2 cells buffer meters apart (somehow) and then check cells count produced by "gridPathCells".
I'm looking for assurence for overlapping, not approximation, with minum cells as possible

**Reply by michaelbrichko** [reply]
Thinking a bit more about it, I think my workarounds aren't good enough, I suspect they will produce too much excess cells.
Let me know please if you think if radius\buffer in meters support in the "horizon" or there is some other good enough workaround

**Reply by nrabinowitz** [reply]
TBH, the best option here is probably to buffer the polygon beforehand and then fill the buffered area. That's what we'd likely do within the library if we wanted to offer a real-distance buffer, but it's much more likely that if we offer a buffer at all it would be a grid-distance buffer. In general, discrete global grids are going to be tricky to work with when you need this kind of exact measurement, because of variations in shape, area, and orientation of grid cells. Better to do the measured transformations on the vector geometry, then move to the grid, rather than the other way around.

**Reply by michaelbrichko** [reply]
Makes sense, thank you!

**Comment by diehl** [comment]
@nrabinowitz If I'm reading the exchange above correctly, we should see the capability to generate a complete hexagonal cover of an arbitrary polygon soon. When do you anticipate this capability will land in `h3-py`? I'm wondering if it's worth waiting for as opposed to developing a workaround in the interim, as I have a need for this capability now. Grateful to see that this is in flight!

**Reply by nrabinowitz** [reply]
@ajfriend should have a better perspective on `h3-py`. The functionality has landed in H3 core but we haven't released it yet.

**Reply by diehl** [reply]
Roger that @nrabinowitz - much appreciated

**Reply by diehl** [reply]
@ajfriend if you have a sense of the ETA for this capability in `h3-py`, I'd love to hear it.

**Reply by eggachecat** [reply]
+1 
why isn't this feature released?

**Reply by ajfriend** [reply]
I think we need to release 4.2 for the core library for this functionality, right? @nrabinowitz
